#!/bin/bash
#SBATCH --account=ctb-sgravel
#SBATCH --cpus-per-task=40
#SBATCH --mem-per-cpu=4G
#SBATCH --time=10:00:00
#SBATCH --output=./log/data_preparation/sbayesr_ld/%j.out
#SBATCH --mail-user=shadi.zabad@mail.mcgill.ca
#SBATCH --mail-type=FAIL


# Input arguments:
extract_file=${1:-"data/keep_files/hq_imputed_variants_hm3.txt"}  # Variants to extract
blocks_file=${2:-"data/ldetect_data/EUR_blocks.bed"}  # Block coordinates
output_dir=${3:-"2_panukb_analysis/model_fit/SBayesRC/data/ld/hapmap3_block"}  # Output directory
genotype_dir=${4:-"data/ukbb_qc_genotypes"}  # Genotype directory
keep_file=${5:-"data/keep_files/ukbb_qc_individuals_EUR.keep"}  # Samples to keep
K=${6:-19}  # Number of parallel tasks
threads=${7:-10}  # Number of threads

# Create the output directory:
mkdir -p "$output_dir"

# ========================================
# Prepare the input data:

module load python/3.11
module load scipy-stack

# Prepare the block coordinates file:
python 2_panukb_analysis/model_fit/SBayesRC/ld/prepare_block_info.py \
    -i "$blocks_file" \
    -o "$output_dir/block_coords.pos"

# Prepare the .ma file containing HapMap3 variants:
python 2_panukb_analysis/model_fit/SBayesRC/ld/prepare_ma_file.py \
    --bim-file "$genotype_dir/chr_*.bim" \
    --extract "$extract_file" \
    --output-file "$output_dir/snp_info.ma"

# Get the number of blocks (exclude the header from the count):
num_blocks=$(tail -n +2 "$output_dir/block_coords.pos" | wc -l)

echo "Number of blocks: $num_blocks"

# If the keep_file is set, then intersect it with the individuals in the genotype data:
if [ -f "$keep_file" ]
then
    python 2_panukb_analysis/model_fit/SBayesRC/ld/merge_keep_file.py \
        --keep "$keep_file" \
        --fam-file "$genotype_dir/chr_1.fam" \
        --output-file "$output_dir/ld_samples.keep"
fi

# ========================================
# Compute the LD matrices:

module load apptainer

echo "> Step 1: Make block LDM info:"

rm -rf "$output_dir/ld_data/"

# Make block LDM info:
apptainer run 2_panukb_analysis/model_fit/SBayesRC/sbayesrc_env/sbayesrc_latest.sif \
    --make-block-ldm-info \
    --block-info "$output_dir/block_coords.pos" \
    --gwas-summary "$output_dir/snp_info.ma" \
    --geno "$genotype_dir/chr_{CHR}" \
    --chr-range 1-22 \
    --out "$output_dir/ld_data" \
    --threads "$threads"


# If the keep_file is set, then update the ld.sh script generated by the previous step
# to explicitly use the keep file (currently not properly supported by the SBayesRC command?):
if [ -f "$keep_file" ]
then
    # Replace all instances of '--bfile' with '--keep $output_dir/ld_samples.keep --bfile':
    sed -i "s#--bfile#--keep $output_dir/ld_samples.keep --bfile#g" "$output_dir/ld_data/ld.sh"

    # Verify the changes:
    head "$output_dir/ld_data/ld.sh"
fi

# Compute the LD matrix:
echo "=============================================================="
echo "> Steps 2-3: Compute LD matrices (& perform eigen decomposition):"

# Loop over the blocks and compute the LD for each block:

compute_transform_block() {
    block_id=$1

    # First task: make-block-ldm
    apptainer run 2_panukb_analysis/model_fit/SBayesRC/sbayesrc_env/sbayesrc_latest.sif \
        --make-block-ldm \
        --block "$block_id" \
        --out "$output_dir/ld_data" \
        --threads 2

    # Second task: make-ldm-eigen
    apptainer run 2_panukb_analysis/model_fit/SBayesRC/sbayesrc_env/sbayesrc_latest.sif \
        --make-ldm-eigen \
        --block "$block_id" \
        --out "$output_dir/ld_data" \
        --threads 2
}

# Export the function and necessary variables
export -f compute_transform_block
export output_dir
export threads

seq 1 $num_blocks | xargs -n 1 -P $K -I {} bash -c 'compute_transform_block "$@"' _ {}

# Merge the data from the LD blocks:
echo "=============================================================="
echo "> Step 4: Merge the data for the LD blocks:"

apptainer run 2_panukb_analysis/model_fit/SBayesRC/sbayesrc_env/sbayesrc_latest.sif \
    --merge-block-ldm-info \
    --out "$output_dir/ld_data" \
    --threads "$threads"
